python/lancedb/db.py

```diff
@@ -47,11 +49,26 @@ def table_names(self) -> list[str]:
         -------
         A list of table names.
         """
-        if get_uri_scheme(self.uri) == "file":
-            return [p.stem for p in Path(self.uri).glob("*.lance")]
-        raise NotImplementedError(
-            "List table_names is only supported for local filesystem for now"
-        )
+        filesystem = None
+        scheme = get_uri_scheme(self.uri)
+        if scheme == "file":
+            filesystem = fs.LocalFileSystem()
+        elif scheme == "s3":
+            filesystem = fs.S3FileSystem()
+        elif scheme == "gs":
+            filesystem = fs.GcsFileSystem()
+        else:
+            raise NotImplementedError(
+                "Unsupported scheme: " + scheme
+            )
```

Comment:
We can let PyArrow get the path for us if we want:

```suggestion
        try:
            filesystem, path = fs.FileSystem.from_uri(self.uri)
        except pa.ArrowInvalid:
            raise NotImplementedError(
                "Unsupported scheme: " + self.uri
            )
```